### PR TITLE
Int 768 victims must be under 18 years old

### PR DESCRIPTION
**Pull request by NagaPradeepGajula-osi**
### Jira Story

- [Victims must be under 18 years old INT-768](https://osi-cwds.atlassian.net/browse/INT-768)

### Purpose
Apply validations for roles based on the dob and approximate age

### CODE REVIEW

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 1]
```diff
@@ -14,13 +15,47 @@ const formatEnums = (enumObject) =>
   Object.keys(enumObject).map((item) => ({label: enumObject[item], value: item}))
 
 export const getPeopleSelector = (state) => state.get('peopleForm')
+export const getScreeningStartDate = (state) => state.getIn(['screeningInformationForm', 'started_at', 'value'])
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: All selectors should end in 'Selector'.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 2]
```diff
@@ -14,13 +15,47 @@ const formatEnums = (enumObject) =>
...
+export const getAgeByDateOfBirth = (state, personId) => {
+  const person = state.getIn(['peopleForm', personId])
+  const dateOfBirth = person.getIn(['date_of_birth', 'value']) || ''
+  if (dateOfBirth) {
+    return moment().diff(dateOfBirth, 'years')
+  }
+  return undefined
+}
+
+export const isUnder18YearsOfAgeAtScreeningDate = (state, personId) => {
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: This math seems overly complicated. Is there any way we can just directly compare the current age to the `started_at` date? That would be a bit more clear than comparing `started_at` to today, and then `date_of_birth` to today, and then those two differences to each other.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 3]
```diff
@@ -14,13 +15,47 @@ const formatEnums = (enumObject) =>
...
+  const approximateAge = person.getIn(['approximate_age', 'value'])
+  const approximateAgeUnit = person.getIn(['approximate_age_units', 'value'])
+  const amountInYears = {
+    years: 1,
+    months: 12,
+    weeks: 52,
+    days: 365,
+  }
+  if (approximateAge) {
+    return approximateAge / amountInYears[approximateAgeUnit]
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: Moment has a built in subtraction function that takes two parameters - an integer and a unit of time. I think we can use that here instead of rolling our own.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 4]
```diff
@@ -29,6 +64,7 @@ export const getErrorsSelector = (state, personId) => {
   const ssn = person.getIn(['ssn', 'value']) || ''
   const ssnWithoutHyphens = ssn.replace(/-|_/g, '')
   return fromJS({
+    roles: combineCompact(isRequiredIfCreate(roles.value, 'Alleged victims must be under 18 years old.', () => (roles.includes('Victim') && !isUnder18YearsOfAgeAtScreeningDate(state, personId)))),
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: This field isn't a required field - it's just a field with an error. Thus, using this function here is somewhat misleading. Is there a way we could rename it to match the current use case? Or don't use it at all and just write a custom validation here?

File: app/javascript/views/people/PersonInformationShow.jsx [hunk 5]
```diff
@@ -31,9 +31,9 @@ const PersonInformationShow = ({
             <ShowField label='Gender'>{gender}</ShowField>
           </div>
           <div className='col-md-5'>
-            <ShowField label='Role(s)'>
-              {roles.length > 0 &&
-                  <ul>{roles.map((role, index) => (<li key={`role-${index}`}>{role}</li>))}</ul>
+            <ShowField label='Role(s)' errors={roles.errors}>
+              {roles.value && roles.value.length > 0 &&
```
Comment (saraOSI) [inline app/javascript/views/people/PersonInformationShow.jsx]: If we need to make sure that `value` is always present, then it should be listed as a required prop or we should set up a default for it. 

However, I have a hunch that this component was failing to render in some of your tests, and you added this to guard agains that. If in a real-world scenario this component will never be passed a "bad" data payload, then the tests should be changed to provide valid inputs, rather than changing the code to work around improper test setup.

File: app/javascript/containers/screenings/PersonDemographicsFormContainer.js [hunk 6]
```diff
@@ -33,7 +33,7 @@ const mergeProps = (stateProps, {dispatch}, {personId}) => {
       {
         dispatch(setField(personId, [field], value))
         dispatch(setField(personId, ['approximate_age'], null))
-        dispatch(setField(personId, ['approximate_age_units'], null))
+        dispatch(setField(personId, ['approximate_age_units'], 'years'))
```
Comment (saraOSI) [inline app/javascript/containers/screenings/PersonDemographicsFormContainer.js]: Why the change in the default value for `approximate_age_units`? If the `date_of_birth` is set, then the `approximate_age` and `approximate_age_units` should both be `null`. 

`null` is a valid value for many of our data points. The code should work around `null` gracefully, not set arbitrary values in an effort to avoid `noFunction` errors.

File: app/javascript/reducers/peopleFormReducer.js [hunk 7]
```diff
@@ -77,7 +77,7 @@ const buildPerson = ({
 }) => fromJS({
   addresses: buildAddresses(addresses),
   approximate_age: {value: approximate_age},
-  approximate_age_units: {value: approximate_age_units},
+  approximate_age_units: {value: approximate_age_units || 'years'},
```
Comment (saraOSI) [inline app/javascript/reducers/peopleFormReducer.js]: Setting a default on page change is bad, but setting a default value in a reducer is _really_ bad. You are changing a value in the data with ZERO interaction from the user to trigger it. `null` is almost always a valid value for our data (or the equivalent of `null`, such as `[]` for an array data type). 

Reducers should not change the data imported from the API. The reducer might changed the way in which the data is stored (the structure of the data), but it should never change the values of the data.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 8]
```diff
@@ -14,13 +15,34 @@ const formatEnums = (enumObject) =>
...
 import US_STATE from 'enums/USState'
 import {RACE_DETAILS} from 'enums/Races'
 import {ETHNICITY_DETAILS} from 'enums/Ethnicity'
 
+export const isUnder18YearsOfAgeAtScreeningDateSelector = (state, personId) => {
+  const screeningStartDate = moment(state.getIn(['screeningInformationForm', 'started_at', 'value']))
+
+  const person = state.getIn(['peopleForm', personId])
+  const dateOfBirth = person.getIn(['date_of_birth', 'value'])
+  const approximateAge = parseInt(person.getIn(['approximate_age', 'value']), 10)
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: How are you avoiding a magic number error here from lint?

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 9]
```diff
@@ -14,13 +15,34 @@ const formatEnums = (enumObject) =>
...
+  const approximateAgeUnit = person.getIn(['approximate_age_units', 'value'])
+
+  let ageFromScreeningDate = false
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: You don't need to instantiate `ageFromScreeningDate` with a value. You can `let` it without assigning it to anything, which means you aren't changing data types in your assignments.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 10]
```diff
@@ -14,13 +15,34 @@ const formatEnums = (enumObject) =>
...
+
+  if (dateOfBirth) {
+    ageFromScreeningDate = screeningStartDate.diff(dateOfBirth, 'years')
+  } else if (approximateAge && approximateAgeUnit) {
+    ageFromScreeningDate = moment().diff(screeningStartDate, 'years') + moment.duration(approximateAge, approximateAgeUnit).asYears()
+  }
+
+  const underAgeLimit = 18
+  return (typeof ageFromScreeningDate === 'number') && ageFromScreeningDate < underAgeLimit
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: If you change `ageFromScreeningDate` above to not instantiate as a boolean, you can simplify this to `ageFromScreeningDate && ageFromScreeningDate < underAgeLimit`

Also, as a total nit pick, I would rename `underAgeLimit` to `victimAgeLimit` or `childAgeLimit` or simply just `ageLimit`.

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 11]
```diff
@@ -31,6 +53,15 @@ export const getErrorsSelector = (state, personId) => {
   return fromJS({
     first_name: combineCompact(isRequiredIfCreate(firstName, 'Please enter a first name.', () => (roles.includes('Victim') || roles.includes('Collateral')))),
     last_name: combineCompact(isRequiredIfCreate(lastName, 'Please enter a last name.', () => (roles.includes('Victim') || roles.includes('Collateral')))),
+    roles: combineCompact(
+      () => {
+        if (roles.includes('Victim') && !isUnder18YearsOfAgeAtScreeningDateSelector(state, personId)) {
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: Is there any benefit to doing a check for `isOver18YearsOfAge`... and not having to pull the inverse of the function?

File: app/javascript/selectors/screening/peopleFormSelectors.js [hunk 12]
```diff
@@ -14,13 +15,34 @@ const formatEnums = (enumObject) =>
...
 export const getPeopleSelector = (state) => state.get('peopleForm')
+
 import {getScreeningIdValueSelector} from 'selectors/screeningSelectors'
 import PHONE_NUMBER_TYPE from 'enums/PhoneNumberType'
 import ADDRESS_TYPE from 'enums/AddressType'
 import US_STATE from 'enums/USState'
 import {RACE_DETAILS} from 'enums/Races'
 import {ETHNICITY_DETAILS} from 'enums/Ethnicity'
 
+export const isUnder18YearsOfAgeAtScreeningDateSelector = (state, personId) => {
```
Comment (saraOSI) [inline app/javascript/selectors/screening/peopleFormSelectors.js]: Is this really a selector, or is it a helper function? Would you expect containers to be able to import it for some use, or is it just here in order to make error checking simpler?

Helper functions are 110% ok in selector files, but they shouldn't be `exported` (they should be treated as private functions), and they shouldn't be called `Selectors`. I think the naming bit might conflict with earlier feedback I gave. My apologies. I didn't read through the code clearly.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 13]
```diff
@@ -11,6 +11,210 @@
...
 
+  context 'Age validation for Roles' do
+    let(:screening) do
+      FactoryGirl.create :screening,
+        participants: [person],
+        started_at: Time.new(2018, 12, 13)
+    end
+
+    context 'Alleged victims age is not under 18 years based on approximate age with unit' do
+      let(:invalid_date_of_birth) { nil }
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: If you're not referencing `invalid_date_of_birth` in this test context, then you can just hardcode it when you create the person on line 27. You only need `let` blocks for variables you want to be able to a) override in future contexts or b) reference in your scenarios.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 14]
```diff
@@ -11,6 +11,210 @@
...
+          approximate_age: invalid_approximate_age,
+          approximate_age_units: invalid_approximate_age_units,
+          roles: ['Victim']
+      end
+      let(:error_messages) do
+        ['Alleged victims must be under 18 years old.']
+      end
+
+      context 'when roles does not include Victim' do
+        let(:person) do
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: You're stomping on the person created at the outside context here. Do you really need the person creation on line 27?

File: spec/features/screening/validations/person_validation_spec.rb [hunk 15]
```diff
@@ -11,6 +11,210 @@
...
+      let(:invalid_approximate_age_units) { 'years' }
+      let(:valid_approximate_age_units) { 'weeks' }
+      let(:person) do
+        FactoryGirl.create :participant,
+          date_of_birth: invalid_date_of_birth,
+          approximate_age: invalid_approximate_age,
+          approximate_age_units: invalid_approximate_age_units,
+          roles: ['Victim']
+      end
+      let(:error_messages) do
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: This probably doesn't need to be an array, since there's only 1 error message.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 16]
```diff
@@ -11,6 +11,210 @@
...
+          within('.card.edit', text: person_name) do
+            fill_in 'Approximate Age', with: valid_approximate_age
+            error_messages.each do |message|
+              expect(page).not_to have_content(message, count: 1)
+            end
+          end
+        end
+      end
+
+      context 'when multiple roles and include victim' do
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: A better context description here might be "when roles includes victim"

File: spec/features/screening/validations/person_validation_spec.rb [hunk 17]
```diff
@@ -11,6 +11,210 @@
     stub_and_visit_edit_screening(screening)
   end
 
+  context 'Age validation for Roles' do
+    let(:screening) do
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: You shouldn't need to stomp on the entire screening for this test context. I think you can just change `started_at` for this context.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 18]
```diff
@@ -11,6 +11,210 @@
...
+            approximate_age: invalid_approximate_age,
+            approximate_age_units: invalid_approximate_age_units,
+            roles: ['Victim', 'Family Member']
+        end
+
+        scenario 'error is displayed until user enters a valid approximate age' do
+          expect(page).to have_content person_name
+          validate_message_as_user_interacts_with_person_card(
+            person: person,
+            error_messages: error_messages,
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: You can pass either `error_message` or `error_messages` to this helper. So, if you change the `let`ted variable to be a single string message, the helper will still accept it.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 19]
```diff
@@ -11,6 +11,210 @@
...
+            person_updates: { roles: [] }
+          ) do
+            within edit_participant_card_selector(person.id) do
+              remove_react_select_option('Role', 'Victim')
+            end
+          end
+        end
+      end
+
+      scenario 'error is displayed until user enters a valid approximate age' do
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: How is this scenario different than the one on line 78?

File: spec/features/screening/validations/person_validation_spec.rb [hunk 20]
```diff
@@ -11,6 +11,210 @@
...
+            person: person,
+            error_messages: error_messages,
+            person_updates: {
+              approximate_age: valid_approximate_age,
+              approximate_age_units: valid_approximate_age_units
+            }
+          ) do
+            within edit_participant_card_selector(person.id) do
+              fill_in 'Approximate Age', with: valid_approximate_age
+              select 'Weeks', from: 'approximate_age_units'
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: `valid_approximate_age_units.capitalize` instead of hardcoding `Weeks` here is likely to be more future-proof.

File: spec/features/screening/validations/person_validation_spec.rb [hunk 21]
```diff
@@ -11,6 +11,210 @@
...
+
+      scenario 'error is displayed until user enters a valid date of birth' do
+        expect(page).to have_content person_name
+        validate_message_as_user_interacts_with_person_card(
+          person: person,
+          error_messages: error_messages,
+          person_updates: { date_of_birth: valid_date_of_birth }
+        ) do
+          within edit_participant_card_selector(person.id) do
+            fill_in_datepicker 'Date of birth', with: Time.new(2002, 12, 13)
```
Comment (saraOSI) [inline spec/features/screening/validations/person_validation_spec.rb]: Why not use the local variable here for `valid_date_of_birth` instead of creating the time object again?

File: spec/javascripts/reducers/peopleFormReducerSpec.js [hunk 22]
```diff
@@ -405,6 +405,16 @@ describe('peopleFormReducer', () => {
         },
       }))
     })
+
+    it('returns people form with approximate age units to be years by default', () => {
```
Comment (saraOSI) [inline spec/javascripts/reducers/peopleFormReducerSpec.js]: Bad News Bears.

File: spec/javascripts/selectors/screening/peopleFormSelectorSpec.js [hunk 23]
```diff
@@ -774,35 +776,108 @@ describe('peopleFormSelectors', () => {
...
     })
+
+    describe('roles in getErrorsSelector', () => {
+      it('returns roles error if role includes Victim and date of birth or approximate age is empty', () => {
+        const peopleForm = {1: {roles: {value: ['Victim'], errors: []}}}
+        const state = fromJS({peopleForm})
+        expect(getErrorsSelector(state, '1').get('roles').first()).toEqual('Alleged victims must be under 18 years old.')
+      })
+
+      it('does not return roles error if role includes Victim and date of birth or approximate age is given', () => {
```
Comment (saraOSI) [inline spec/javascripts/selectors/screening/peopleFormSelectorSpec.js]: *valid date of birth is given

File: spec/features/screening/validations/person_validation_spec.rb (hunk 19 above)
Comment (NagaPradeepGajula-osi) [reply spec/features/screening/validations/person_validation_spec.rb]: the context for these similar scenarios in 78 and here are different.

File: spec/javascripts/selectors/screening/peopleFormSelectorSpec.js [hunk 24]
```diff
@@ -774,35 +775,74 @@ describe('peopleFormSelectors', () => {
...
       expect(getErrorsSelector(state, '1').get('first_name').first()).toEqual(undefined)
     })
+
     it('returns undefined if last name is empty and role does not include Victim', () => {
       const peopleForm = {1: {roles: {value: ['Some role']}}}
       const state = fromJS({peopleForm})
       expect(getErrorsSelector(state, '1').get('last_name').first()).toEqual(undefined)
     })
+
+    describe('roles in getErrorsSelector', () => {
```
Comment (saraOSI) [inline spec/javascripts/selectors/screening/peopleFormSelectorSpec.js]: The tests you had for deeply unit testing the selector you made private should probably be here, yes?

Comment (NagaPradeepGajula-osi) [reply spec/javascripts/selectors/screening/peopleFormSelectorSpec.js]: added under roles error selector test cases
